[WebAuthn] Trim pubKeyCredParams if possible

pascoej opened the pull request:
#### 55a705b37c9e6713d96fe5b51a9d2dfc775c2e44
<pre>
[WebAuthn] Trim pubKeyCredParams if possible
<a href="https://bugs.webkit.org/show_bug.cgi?id=283069">https://bugs.webkit.org/show_bug.cgi?id=283069</a>
<a href="https://rdar.apple.com/problem/139819203">rdar://problem/139819203</a>

Reviewed by Brent Fulgham.

As noted in <a href="https://bugs.webkit.org/show_bug.cgi?id=220415">https://bugs.webkit.org/show_bug.cgi?id=220415</a>, some RPs request every
algorithm they support. This can cause make credential requests to hit maxMsgSize
whenever it is sufficently low. In most cases, we know the algorithms a security
key supports via it&apos;s getInfo response, so we can trim down the algorithms we actually
send to the authenticator to keep the message size small.

This patch does that by looking at the algorithms supported in getInfo and the algorithms
requested by the RP. The list of algorithms requested by the RP is sorted in order of preference,
if we know the security key supports one of them, we can just send that one. If we know the
security key supports none of the requested algorithms, we can trim down the list sent to the
key to just one. If the security key does not tell us what it supports, then we can&apos;t trim the
list.

Added API tests to exercise this behavior.

* Source/WebCore/Modules/webauthn/fido/AuthenticatorGetInfoResponse.h:
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
(fido::trimmedParameters):
(fido::encodeMakeCredentialRequestAsCBOR):
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(+[_WKWebAuthenticationPanel encodeMakeCredentialCommandWithClientDataHash:options:userVerificationAvailability:authenticatorSupportedCredentialParameters:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanelForTesting.h:
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::makeCredential):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKWebAuthenticationPanel.mm:
(TestWebKitAPI::TEST(WebAuthenticationPanel, EncodeCTAPCreationTrimmedParametersGetInfoNoneES256)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, EncodeCTAPCreationTrimmedParametersGetInfoNoneRS256)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, EncodeCTAPCreationTrimmedParametersGetInfoSupportsEC2)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, EncodeCTAPCreationTrimmedParametersGetInfoSupportsDisjoint)):

Canonical link: <a href="https://commits.webkit.org/295258@main">https://commits.webkit.org/295258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ae4b5c3ac7ebeb10036cfc8291d1dbbd06e2c46

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104586 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24296 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14714 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109798 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/55257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106626 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24682 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32841 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/79417 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/55257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107592 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19194 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94391 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59729 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12466 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54630 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/12518 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112188 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31747 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88497 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32111 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90625 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/88119 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22438 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33020 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10789 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/27064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31674 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37015 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31466 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34805 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33026 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
```diff
@@ -88,13 +89,41 @@ static CBORValue convertDescriptorToCBOR(const PublicKeyCredentialDescriptor& de
+        if (parameter.type != PublicKeyCredentialType::PublicKey)
+            continue;
+        if (authenticatorSupportedAlgorithms.contains(parameter.alg))
+            return { parameter };
```
So this always returns the first option supported by the authenticator. Does that cause any problem if there is a "preferred" algorithm that happens to fall later in the list of choices?

pascoej replied:
The list is ordered from most preferred to least preferred. 

> This sequence is ordered from most preferred (by the RP) to least preferred.
from https://fidoalliance.org/specs/fido-v2.0-ps-20190130/fido-client-to-authenticator-protocol-v2.0-ps-20190130.html

brentfulgham commented on Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
```diff
@@ -88,13 +89,41 @@ static CBORValue convertDescriptorToCBOR(const PublicKeyCredentialDescriptor& de
+        // Support for ES256 required by U2F backwards compatibility.
+        // https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-client-to-authenticator-protocol-v2.0-id-20180227.html#u2f-authenticatorMakeCredential-interoperability
+        if (parameter.alg == COSE::ES256)
+            return { parameter };
```
So if the request includes `COSE::ES256`, we just use it and don't return any other options?

pascoej replied:
We use it if it's included in the list of algorithms requested by the RP. The list provided by the RP is most preferred to least preferred, so if we reach here we know the RP wants ES256 and the authenticator will support it.

brentfulgham commented on Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
```diff
@@ -88,13 +89,41 @@ static CBORValue convertDescriptorToCBOR(const PublicKeyCredentialDescriptor& de
+
+    // We know the algorithms the authenticator supports and none of those were requested.
+    if (authenticatorSupportedAlgorithms.size())
+        return { parameters.first() };
```
Why is it better to return just the first unsupported algorithm, rather than the entire set? I don't really understand why this is better for a case where we can't use any of the parameters.

pascoej replied:
The issue is that some websites will ask for every last algorithm they support in the list they provide, which can cause the message sent for makeCredential to become too large to be sent over a transport like NFC. The idea with this patch is that we know the processing steps the authenticator will take from the CTAP spec and the authenticator already reports what algorithms it supports in its getInfo response.

Using that we can prune the list to avoid hitting maxMsgSize whenever the RP sends a long list.

For example, apparently one RP sends a request like:
```{ "attestation": "direct", "authenticatorSelection": { "residentKey": "discouraged", "userVerification": "preferred" }, "challenge": "a0oPIQ4k5mF15rK1WKnUxqGsRYvsC-0tHf3er8i0LZHTv8f4KGUMRQ1ys- aJiFg0_OuIPhKA8ecaRuUmfHN3EQ", "excludeCredentials": [ {"id": "iNRyobHFV_khWBgdUXpkeNxVFjRh2MIZI7YSJt4qBDTsK1kwQ8H6DmqixmZM2jRIbhXpbSV_UN_9GK4PcuMYv4EMReqzRhIPAgggCK21ouy4fHE HUFHHAYdMTC6d8Q", "transports": [ "usb", "nfc", "ble", "hybrid", "internal" ], "type": "public-key" }, { "id": "vxwsDOgJ08syZrS13YAEXS7UV51B0rb8гBTm92NJvWaxMEQOUkc2L_Nvs Se6Cprd@VG5Wt-0WUcuDx5xdomaIA", "transports": [ "usb", "nfc", "ble", "hybrid", "internal" ], "type": "public-key" } ], "extensions": { "credProps": true }, "pubKeyCredParams": [ { "alg":-65535, "type": "public-key" }, { "alg":-257, "type": "public-key" }, { "alg": -258, "type": "public-key" }, { "alg": -259, "type": "public-key" }, { "alg": -37, "type": "public-key" }, { "alg": -38, "type": "public-key" }, { "alg":-39, "type": "public-key" }, { "alg": 7, "type": "public-key" }, { "alg":-35, "type": "public-key" }, { "alg": -36, "type": "public-key" }, { "alg": -8, "type": "public-key" }, { "alg": -43, "type": "public-key" } ], "rp": { "id": "...", "name": "..." }, "user": { "displayName": "Chrome V126.0.0.0 (Windows)", "id": "NDg2MTA4NDIz", "name": "[xxxxxx.user1@gmail.com](mailto:xxxxxx.user1@gmail.com)" } }```

Sending all those algorithms can easily exceed the maxMsgSize for NFC, especially because "public-key" is encoded as a string in the CTAP spec.

brentfulgham replied:
It's not surprising that returning one response is a smaller message than all possible things. I'm just unclear why returning the first thing (highest preference) algorithm our authenticator supports when we don't match any of the things the website asked for is appropriate. Shouldn't we have some kind of error state or something?

pascoej replied:
My logic here was just to maintain the same behavior wrt to the authenticator with the original list vs the truncated list. In both cases the authenticator will return CTAP2_ERR_UNSUPPORTED_ALGORITHM and our existing error handling code will handle that. I'd like to avoid creating two separate paths for this error to get handled.

You bring up a good point about the error handling here. Right now we actually do the wrong thing by not handling errors like `CTAP2_ERR_UNSUPPORTED_ALGORITHM` (and maybe`CTAP2_ERR_INVALID_OPTION `, `CTAP2_ERR_UNSUPPORTED_OPTION`) returned from the authenticator.

These errors are returned before an authorization gesture is collected by the authenticator. It doesn't makes sense to report the error right after we get it back from the makeCredential request, because collecting an authorization gesture is the method used to select an authenticator to use for the ceremony.

For example, if the user has 3 security keys plugged in (or an unsupported security key plugged in, but another that is supported that they plug in during the ceremony,) it doesn't make sense to report the error whenever we get it back from makeCredential because we don't know if the authenticator that reported the error is the one the user wants to use for the ceremony. 

Instead we should note down we get this error from a particular authenticator and submit a dummy request to the authenticator to collect a tap. If they do tap the authenticator where we got this error, we can show a message that this authenticator cannot be used because it doesn't support an algorithm the RP wants.

brentfulgham commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1211,6 +1211,19 @@ + (NSData *)encodeMakeCredentialCommandWithClientDataHash:(NSData *)clientDataHa
     return encodedCommand.autorelease();
 }
 
++ (NSData *) encodeMakeCredentialCommandWithClientDataHash:(NSData *)clientDataHash options:(_WKPublicKeyCredentialCreationOptions *)options userVerificationAvailability:(_WKWebAuthenticationUserVerificationAvailability)userVerificationAvailability authenticatorSupportedCredentialParameters:(NSArray<_WKPublicKeyCredentialParameters *> *)credentialParameters
```
This appears to be solely for testing, is that right?

pascoej replied:
Yes, we don't have another way to check the generated CTAP request bytes in our testing framework right now.

brentfulgham reviewed (approved):
I'm still not quite sure why returning the highest priority supported algorithm is right when we don't support any of the items the site asked for. But otherwise this seems fine.

pascoej committed:
[WebAuthn] Trim pubKeyCredParams if possible
https://bugs.webkit.org/show_bug.cgi?id=283069
rdar://problem/139819203

Reviewed by Brent Fulgham.

As noted in https://bugs.webkit.org/show_bug.cgi?id=220415, some RPs request every
algorithm they support. This can cause make credential requests to hit maxMsgSize
whenever it is sufficently low. In most cases, we know the algorithms a security
key supports via it's getInfo response, so we can trim down the algorithms we actually
send to the authenticator to keep the message size small.

This patch does that by looking at the algorithms supported in getInfo and the algorithms
requested by the RP. The list of algorithms requested by the RP is sorted in order of preference,
if we know the security key supports one of them, we can just send that one. If we know the
security key supports none of the requested algorithms, we can trim down the list sent to the
key to just one. If the security key does not tell us what it supports, then we can't trim the
list.

Added API tests to exercise this behavior.

* Source/WebCore/Modules/webauthn/fido/AuthenticatorGetInfoResponse.h:
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
(fido::trimmedParameters):
(fido::encodeMakeCredentialRequestAsCBOR):
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(+[_WKWebAuthenticationPanel encodeMakeCredentialCommandWithClientDataHash:options:userVerificationAvailability:authenticatorSupportedCredentialParameters:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanelForTesting.h:
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::makeCredential):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKWebAuthenticationPanel.mm:
(TestWebKitAPI::TEST(WebAuthenticationPanel, EncodeCTAPCreationTrimmedParametersGetInfoNoneES256)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, EncodeCTAPCreationTrimmedParametersGetInfoNoneRS256)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, EncodeCTAPCreationTrimmedParametersGetInfoSupportsEC2)):
(TestWebKitAPI::TEST(WebAuthenticationPanel, EncodeCTAPCreationTrimmedParametersGetInfoSupportsDisjoint)):

Canonical link: https://commits.webkit.org/295258@main
